### ISSUE TITLE
cutIntroG throws exception on expansion trees (which work fine for cutIntro)

### ISSUE THREAD

**Issue by newca12** [newca12/gapt#250]
_From [stefan.hetzl](https://code.google.com/u/stefan.hetzl/) on November 20, 2013 09:23:05_

Consider the following CLI-commands:

val E = loadVeriTProof( "../testing/veriT-SMT-LIB/QF_UF/eq_diamond/eq_diamond2.smt2.proof_flat" )
val q1 = cutIntroG( E, NoConstraint )

This leads to an exception being thrown from Deltas.scala. On the other hand, the commands

val q2 = cutIntro( E )
val q3 = cutIntroExp( E )

lead to sucessful cut-introductions.

_Original issue: http://code.google.com/p/gapt/issues/detail?id=250_

**Comment by newca12** [comment]
_From [janos.ta...@gmail.com](https://code.google.com/u/110287254390562270748/) on November 21, 2013 16:23:12_

Revision `#1769` fixes this. The failure was due to FOLVars in the termset, which got mixed up with the eigenvariables.

A related issue is name clashes: VeriT-output is probably unproblematic in this regard, but cutIntroG will fail if an expansion tree contains any of the following as free variables: α_n, x, x_n, y, z, u, v, w (n>=0) (in the case of cutIntro, α and x are similarly fatal).
I don't know how pressing this is, but for the time being, I renamed the bound variables of the cut formula to β_n (n>=0), since it's unlikely that many programs will use unicode variable names. Still, the vulnerability, in principle, exists.

Should I revert this and go back to naming the CF's variables x,y,z,..., or should we find a general solution, like renaming all free variables in a preprocessing step, or identifying eigenvariables explicitly by, say, creating a subclass of FOLVar?

**Comment by newca12** [comment]
_From [stefan.hetzl](https://code.google.com/u/stefan.hetzl/) on December 04, 2013 02:19:51_

The point is that expansion trees that are used as input for cut-introduction do not (i.e. in theory) contain free variables. The bound variables of cutIntroG should be named x_1,x_2,... Changing this back should work now after r1801 which correctly parses the individual symbols of veriT as constants.

**Cc:** giselle....@gmail.com

**Comment by newca12** [comment]
_From [stefan.hetzl](https://code.google.com/u/stefan.hetzl/) on December 12, 2013 09:01:15_

I have now changed the names of bound variables back to x_0,x_1,... which are more common names for bound variables; \alpha, \beta,... are more commonly used for eigenvariables.

The example mentioned at the start of this issue works now because veriT-import has been fixed to not create variables anymore. I am therefore closing this issue.

The general problem of what to do if the expansion proof input to cut-introduction contains variables is not solved. A good long-term solution would be to do cut-introduction with a blacklist of forbidden variables (those that appear in the original input).

**Status:** Fixed
